vo_gpu: implement error diffusion for dithering

bjin opened the pull request:
This is a straightforward parallel implementation of error diffusion
algorithms in compute shader. Basically we use single work group with
maximal possible size to process the whole image. After a shift
mapping we are able to process all pixels column by column.

A large ring buffer are allocated in shared memory to speed things up.
However the size of required shared memory depends linearly on the
height of video window (or screen height in fullscreen mode). In case
there is no enough share memory, it will fallback to '--dither=fruit'.

The maximal allowed work group size is hardcoded as 1024. Ideally we
could query GL_MAX_COMPUTE_WORK_GROUP_INVOCATIONS. But for whatever
reason, it seems most high end card from nvidia and amd support only
the minimal required value, so I guess we can stick to it for now.

Tested on Linux with Intel HD 4000. Rendering time is ~~22ms~~ 14ms on 1920x1080 for `sierra-lite`.

bjin commented:
Some samples with `--vf format=gray --dither-depth=1`

* `--dither=error-diffusion --error-diffusion=sierra-lite`:
![sierra-lite](https://user-images.githubusercontent.com/175336/54474620-0024a400-4822-11e9-9ce5-0c33db1f404c.png)
* `--dither=error-diffusion --error-diffusion=floyd-steinberg`:
![floyd-steinberg](https://user-images.githubusercontent.com/175336/54474618-ff8c0d80-4821-11e9-8035-2bbb9df11284.png)
* `--dither=error-diffusion --error-diffusion=jarvis-judice-ninke`:
![jarvis-judice-ninke](https://user-images.githubusercontent.com/175336/54474616-fef37700-4821-11e9-901f-a92f26cde284.png)
* `--dither=error-diffusion --error-diffusion=atkinson`:
![atkinson](https://user-images.githubusercontent.com/175336/54474617-ff8c0d80-4821-11e9-8451-6f9409075bc1.png)
* `--dither=error-diffusion --error-diffusion=simple`:
![simple](https://user-images.githubusercontent.com/175336/54474619-0024a400-4822-11e9-93fe-0de6df85a6a3.png)
* `--dither=fruit`:
![fruit](https://user-images.githubusercontent.com/175336/54474622-00bd3a80-4822-11e9-8e54-a73ce38390b3.png)
* `--dither=ordered`:
![ordered](https://user-images.githubusercontent.com/175336/54474621-00bd3a80-4822-11e9-875b-f449a15282c8.png)

haasn commented:
.....why?

haasn commented:
I mean this is very impressive work, but at 22ms per frame, what is the purpose of having this in a multimedia player? Also, how well does it work on actual videos? Aren't error diffusion algorithms infamous for "jittering" around a lot in result to small fluctuations in the input?

bjin commented:
> 22ms per frame

Note that it's an old Intel GPU back to like six years ago. I expect rendering time to be much shorter on amd and nvidia card.

> what is the purpose of having this in a multimedia player?

Even if it's not possible to run error diffusion in real time, it can still be used to process screenshots. There is case where I want to save the screenshot with best dithering option available (along with mpv's color management). Now it's possible to run it in real-time, I didn't see a reason not to do so. This is probably about how placebo setting is: run the best options your GPU can handle without frame drop, and forget about bothering.

> Aren't error diffusion algorithms infamous for "jittering" around a lot in result to small fluctuations in the input?

This is exactly what `--temporal-dither` introduces to `--dither=fruit`. It's basically invisible for high bit-depth dithering anyway.

haasn commented:
I understand. For an image viewer you are right, it makes more sense. And for some reason, good dithering is annoyingly difficult to accomplish with other methods (e.g. imagemagick).

The code is isolated enough that I'd be happy merging it as long as it doesn't get into our way in the future. I will provide some code comments though, when I get the time to review this properly.

haasn commented:
For starters, why are you using a string option for the dither method instead of an `enum` + `OPT_CHOICE` approach?

bjin commented:
> For starters, why are you using a string option for the dither method instead of an enum + OPT_CHOICE approach?

The error diffusion kernels are defined in its own ".c" like filter kernels. enum for each filter seems rather annoying and unnecessary to me. In the future, all it takes to add new kernel is just adding a new definition in the constant array in its own file, without messing the huge "video.c".

bjin commented:
I added another commit which lowered the shared memory requirement and the rendering time considerably. It's now 14ms for `sierra-lite` and 33ms for `jarvis-judice-ninke` (one of the slowest kernel, can't handle this before due to shared memory requirement). However, due to loss of precision during error propagation (8 bit instead of 24 bit before), the result is not bit-exact as before.

* `jarvis-judice-ninke` before
![jarvis-judice-ninke](https://user-images.githubusercontent.com/175336/54487107-b80e8b80-48cc-11e9-985b-3b04b4bc1c7c.png)

* `jarvis-judice-ninke` with the new change
![jarvis-judice-ninke-new](https://user-images.githubusercontent.com/175336/54528297-86beba00-49b7-11e9-98cc-e0454cc8db81.png)

bjin commented:
ping @haasn

bjin commented:
For record purpose, here is the list of changes since the very first revision
* fixed a visible artifacts near left border
* used different texture for source and rendering target for error diffusion, works with `gpu-api=vulkan` now
* changed to use single uint instead of three for storing errors, saved shared memory usage and improved rendering performance considerably
* various comments and documents improvement

bjin commented:
@haasn another friendly ping.

Please tell if there are any action required from me to make the review/merging easier for you (like preparing a patch for libplacebo as well?)

haasn commented on video/out/gpu/video.c:
```diff
@@ -3524,7 +3579,8 @@ static bool check_dumb_mode(struct gl_video *p)
         o->linear_upscaling || o->sigmoid_upscaling || o->interpolation ||
-        o->blend_subs || o->deband || o->unsharp)
+        o->blend_subs || o->deband || o->unsharp ||
+        o->dither_algo == DITHER_ERROR_DIFFUSION)
```
Error diffusion probably works in dumb mode as well

bjin replied:
It probably wont. It requires compute shader capability, and an intermediate FBO texture. Both are not available in dumb mode. The changes to `pass_dither()` actually called `finish_pass_tex()` twice for such purpose.

haasn replied:
They're still usable in "voluntary dumb mode". The only time you need to force-disabled this feature is if compute shaders / FBOs are unavailable, but that check already happens elsewhere.

bjin replied:
okay, I understand your point now. pushed another commit to enable error diffusion in dumb mode.

haasn commented on video/out/gpu/error_diffusion.c:
```diff
@@ -141,54 +136,85 @@ void pass_error_diffusion(struct gl_shader_cache *sc,
+    //     +------------------------------------+
+    //     |RRRRRRRR|0000|GGGGGGGG|0000|BBBBBBBB|
+    //     +------------------------------------+
+    //
```
What happens when we overflow addition on the `R` component?

bjin replied:
The high bit will be discarded, if there is an overflow after add operation. This is actually intended, to have operations on signed and unsigned integers align to each other.

    (-1)+(-1)=0xffffffff+0xffffffff=0xfffffffe=-2
    (-1)*(-1)=0xffffffff+0xffffffff=0x00000001=1

haasn commented:
Gave it some glances. I haven't looked at or thought about the algorithm itself too much, but the code looks reasonable otherwise. Since it's a "fringe" feature I'd probably be happy merging this as-is but see my comments.

bjin commented:
Really appreciate for your review. Please see my reply addressing your comments.

bjin commented:
rebased to master. updated code comment to address one of your comment

haasn commented:
Seems okay to me with the changes. Untested

bjin commented:
Changes since last approval:

1. rebased, squashed fixes commits. So this PR can be merged as-is now
2. Fixed flaky use of intermediate textures, use new standalone textures for all rendering. This should solve any flaky issues on vulkan now.

@wsldankers also reported an issue of this PR on IRC. It somehow caused distortion when used along with compute shader EWA scaler on NVIDIA GPU. AFAICT now, it's most likely a NVIDIA issue. I definitely want to get this fixed or find a proper workaround, but it's unlikely to happen anytime soon since I don't have NVIDIA GPU to test.

Due to natural of this PR (a "fringe" feature), and also there is an easy workaround for the reported issue (alternative scaler). I wouldn't consider it as a blocker of this PR. But probably we should add this as a known issue to the manual. Or just leave it in this PR thread, for anyone following this thread closely.

wsldankers commented:
Unable to reproduce with newer proprietary nvidia driver (430). I'd say case closed, worrying over proprietary driver bugs is a waste of time.

bjin commented:
split out the newest change into a single commit, again, for reviewing purpose. (Please squash the first two commits before merging).

sergeevabc commented:
Is it correct to assume that it’s better to switch dithering off completely by means of ```dither-depth=no``` if I need to capture video frame as is, no matter the final file size?

haasn commented:
If you need to capture the video frame "as is", you also need to disable RGB conversion and chroma upscaling.
